source/server/api_listener_impl.h

```diff
@@ -0,0 +1,198 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/config/core/v3alpha/base.pb.h"
+#include "envoy/config/listener/v3alpha/listener.pb.h"
+#include "envoy/network/connection.h"
+#include "envoy/network/filter.h"
+#include "envoy/server/api_listener.h"
+#include "envoy/server/drain_manager.h"
+#include "envoy/server/filter_config.h"
+#include "envoy/server/listener_manager.h"
+#include "envoy/stats/scope.h"
+
+#include "common/common/empty_string.h"
+#include "common/common/logger.h"
+#include "common/http/conn_manager_impl.h"
+#include "common/init/manager_impl.h"
+#include "common/stream_info/stream_info_impl.h"
+
+namespace Envoy {
+namespace Server {
+
+class ListenerManagerImpl;
+
+/**
+ * Listener that provides a handle to inject HTTP calls into envoy via an Http::ConnectionManager.
+ * Thus it provides full access to Envoy's L7 features, e.g HTTP filters.
+ */
+class ApiListenerImpl : public ApiListener,
+                        public Configuration::FactoryContext,
+                        public Network::DrainDecision,
+                        Logger::Loggable<Logger::Id::http> {
+public:
+  // TODO(junr03): consider moving Envoy Mobile's SyntheticAddressImpl to Envoy in order to return
+  // that rather than this semi-real one.
+  const Network::Address::InstanceConstSharedPtr& address() const { return address_; }
+
+  // ApiListener
+  absl::string_view name() const override { return name_; }
+
+  // TODO(junr03): move these functions into a concrete Server:Configuration::FactoryContext that
+  // takes the server's context, and the listener config in the constructor. The only challenging
+  // part is implementing the initManager() call. The function depends on the Listener's state for
+  // ListenerImpl, and will do so once ApiListeners are dynamic.
```

Comment:
Can we just do this now with the utility class that @lambdai added? It's already basically a wrapper and it seems not difficult to share the wrapper in some way with this code.

Reply:
I went ahead and created an object here. I don't think there is a clear way to use the wrapper pattern and not have the listener implement the FactoryContext interface, because someone has to implement that interface. i.e, someone has to be the inner wrapped object. Let me know if you see an alternative I am missing here.

Reply:
cc @lambdai if you want to take a look

Reply:
To me the FilterChainContext should delegate functionality to another (shared) object to avoid duplicating the code/members in replicas of FilterChainContext.
My understanding the shared object is listener_impl/api_listener_impl now, and could be future wrapped thing if we unify some code between plain listener_impl and api_listener_impl

Reply:
And the unify could be an independent PR, as long as the impl of api listener is not immediately abused by elsewhere